internal/ecgit/ecgit.go

```diff
@@ -0,0 +1,306 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+package ecgit
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"os"
+	"os/exec"
+	"strings"
+	"time"
+
+	"github.com/go-git/go-git/v5"
+	"github.com/go-git/go-git/v5/plumbing"
+	"github.com/go-git/go-git/v5/plumbing/object"
+	"github.com/go-git/go-git/v5/plumbing/transport/http"
+	"github.com/google/go-github/v45/github"
+	log "github.com/sirupsen/logrus"
+	"github.com/spf13/afero"
+	giturls "github.com/whilp/git-urls"
+	"golang.org/x/oauth2"
+
+	"github.com/hacbs-contract/ec-cli/internal/utils"
+)
+
+var CloneRepoWithAuth = cloneRepoWithAuth
+var CreateAutomatedPR = createAutomatedPR
+var CreateBranch = createBranch
+var CreatePullRequest = createPullRequest
+var ApplyDiff = applyDiff
+var CommitChange = commitChange
+var PushChange = pushChange
+var NewGithubClient = newGithubClient
+
+// cloneRepoWithAuth clones the specified repository.
+func cloneRepoWithAuth(repoURL string, destDir string, ctx context.Context) (*git.Repository, error) {
+	log.Info("No private key file provided, using default authentication")
+	if os.Getenv("GITHUB_USERNAME") == "" {
+		log.Debug("Unable to clone repository, GITHUB_USERNAME not set")
+		return nil, errors.New("GITHUB_USERNAME not set")
+	}
+	if os.Getenv("GITHUB_TOKEN") == "" {
+		return nil, errors.New("GITHUB_TOKEN not set")
+	}
+	auth := &http.BasicAuth{
+		Username: os.Getenv("GITHUB_USERNAME"),
+		Password: os.Getenv("GITHUB_TOKEN"),
+	}
+
+	r, err := git.PlainCloneContext(ctx, destDir, false, &git.CloneOptions{
+		Auth:     auth,
+		URL:      repoURL,
+		Progress: os.Stdout,
+	})
+	if err != nil {
+		log.Debug("Unable to clone repository due to error")
+		return nil, err
+	}
+	return r, nil
+}
+
+// createBranch creates a new branch on the specified repository.
+func createBranch(repo *git.Repository, branchName string) (*git.Worktree, error) {
+	// Create a new branch
+	headRef, err := repo.Head()
+	if err != nil {
+		log.Debug("Unable to create branch due to getting HEAD ref")
+		return nil, err
+	}
+	refName := plumbing.NewBranchReferenceName(branchName)
+	ref := plumbing.NewHashReference(refName, headRef.Hash())
+	err = repo.Storer.SetReference(ref)
+	if err != nil {
+		log.Debugf("Error setting ref: %s\n", ref)
+		return nil, err
+	}
+	w, err := repo.Worktree()
+	if err != nil {
+		log.Debug("Unable to get repo worktree")
+		return nil, err
+	}
+	err = w.Checkout(&git.CheckoutOptions{
```

Comment:
I'm surprised this actually creates a branch! Is `CheckoutOptions{Create: true}` not needed?

Reply:
Seemingly no.